core/src/main/java/io/grpc/transport/Http2ClientStream.java

```diff
@@ -113,30 +114,31 @@ protected void transportHeadersReceived(Metadata.Headers headers) {
    * @param frame the received data frame
    * @param endOfStream {@code true} if there will be no more data received for this stream
    */
-  protected void transportDataReceived(ReadableBuffer frame, boolean endOfStream) {
+  protected final void transportDataReceived(ReadableBuffer frame, boolean endOfStream) {
     if (transportError == null && inboundPhase() == Phase.HEADERS) {
       // Must receive headers prior to receiving any payload as we use headers to check for
       // protocol correctness.
       transportError = Status.INTERNAL.withDescription("no headers received prior to data");
     }
-    if (transportError != null) {
-      // We've already detected a transport error and now we're just accumulating more detail
-      // for it.
-      transportError = transportError.augmentDescription("DATA-----------------------------\n"
-          + ReadableBuffers.readAsString(frame, errorCharset));
-      frame.close();
-      if (transportError.getDescription().length() > 1000 || endOfStream) {
-        inboundTransportError(transportError);
-        // We have enough error detail so lets cancel.
-        sendCancel(Status.CANCELLED);
-      }
-    } else {
+
+    if (transportError == null) {
       inboundDataReceived(frame);
       if (endOfStream) {
         // This is a protocol violation as we expect to receive trailers.
         transportError = Status.INTERNAL.withDescription("Recevied EOS on DATA frame");
         inboundTransportError(transportError);
       }
+      return;
+    }
+    // We've already detected a transport error and now we're just accumulating more detail
```

Comment:
Hmm, my hope was to move the common case to be the first thing in the method (thus switching the transport error check to look for null rather than not null).